test/inputs-dynamodb-data-test.js

```diff
@@ -0,0 +1,305 @@
+
+      expect(records.length).to.equal(2);
+      expect(records[0]).to.eql({
+        type: 'postbytes',
+        listenerEpisode: 'le',
+        digest: 'd',
+        timestamp: 1000,
+        some: 'data',
+      });
+      expect(records[1]).to.eql({
+        type: 'postbytes',
+        listenerEpisode: 'le',
+        digest: 'd',
+        timestamp: 100000,
+        some: 'data',
+      });
+    });
+
+    it('filters downloads', () => {
+      const download = { the: 'download' };
+      const segments = {
+        1000: true,
+        100000: false,
+        100000.1: true,
+        200000: true,
+        200000.2: true,
+      };
+      const records = ddb.format(['le.d', { download }, segments]);
+
+      expect(records.length).to.equal(3);
+      expect(records[0].download).to.eql(download);
+      expect(records[1].download).to.be.undefined;
+      expect(records[2].download).to.eql(download);
+    });
+
+    it('filters impressions', () => {
+      const impressions = [
+        { segment: 0, num: 'imp0' },
+        { segment: 3, num: 'imp3' },
+        { segment: 2, num: 'imp2' },
+      ];
+      const segments = {
+        1000: true,
+        100000.2: true,
+        '100001.0': false,
+        '100002.0': true,
+        '200000.0': true,
+        200001.2: false,
+        200002.3: true,
+      };
+      const records = ddb.format(['le.d', { impressions }, segments]);
+
+      expect(records.length).to.equal(3);
+      expect(records[0].timestamp).to.equal(1000);
+      expect(records[0].impressions).to.eql([]);
+      expect(records[1].timestamp).to.equal(100000);
+      expect(records[1].impressions).to.eql([impressions[2]]);
+      expect(records[2].timestamp).to.equal(200002);
+      expect(records[2].impressions).to.eql([impressions[0], impressions[1]]);
+    });
+  });
+
+  describe('#dedupSegments', () => {
+    const ddb = new DynamodbData();
+
+    it('filters old segments', () => {
+      const [segments] = ddb.dedupSegments({
+        1000: true,
+        1000.1: false,
+        1000.2: false,
+        1000.3: true,
+        1000.4: true,
+      });
+      expect(segments).to.eql({ 19700101: ['DOWNLOAD', 3, 4] });
+    });
+
+    it('returns the max timestamp per segment per day', () => {
+      const [, timestamps] = ddb.dedupSegments({
+        1000: true,
+        3000.1: true,
+        5000.2: true,
+        2000.3: true,
+        1000.4: true,
+      });
+      expect(timestamps).to.eql({ 19700101: 5000 });
+    });
+
+    it('dedups segments by utc day', () => {
+      const [segments, timestamps] = ddb.dedupSegments({
+        1000: true,
+        2000: false,
+        2000.4: false,
+        100000: true,
+        100000.2: false,
+        100001.2: true,
+        200001.2: true,
+        200000.4: true,
+      });
+      expect(segments).to.eql({ 19700102: ['DOWNLOAD'], 19700103: [2, 4] });
+      expect(timestamps).to.eql({ 19700102: 100000, 19700103: 200001 });
```

Comment:
Curious, why isn't there a `19700101` key in here?

Reply:
In this case, there are 3 `segments` in DDB falling on 1970-01-01:

```
{
  1000: true,
  2000: false,
  2000.4: false,
  ...
}
```

So this indicates DDB already contained 2 segments at 2000ms into the day, for the overall-download and segment `4`.

Since `1000` is true, we just now `ddb.updateItem'd` a new overall-download at 1000ms into the day.  But since we already counted 1970-01-01 ... it gets dedup'd.